File: backend/src/main/java/com/birdkingdom/entity/ForumPost.java
package com.birdkingdom.entity;

import jakarta.persistence.*;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "forum_posts")
public class ForumPost {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "author_id", nullable = false)
    private User author;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String content;

    @Column(name = "post_type", length = 20)
    private String postType = "NORMAL";
    
    @Column(name = "media_type", length = 20)
    private String mediaType = "IMAGE"; // IMAGE, VIDEO
    
    @Column(name = "video_url", length = 500)
    private String videoUrl;
    
    @Column(name = "video_cover", length = 500)
    private String videoCover;
    
    @Column(name = "video_duration")
    private Integer videoDuration; // 视频时长（秒）

    @Column(name = "like_count")
    private Integer likeCount = 0;

    @Column(name = "comment_count")
    private Integer commentCount = 0;

    @Column(name = "view_count")
    private Integer viewCount = 0;

    @Column(precision = 10, scale = 7)
    private BigDecimal latitude;

    @Column(precision = 10, scale = 7)
    private BigDecimal longitude;

    @Column(name = "location_name", length = 100)
    private String locationName;

    @Column(name = "bird_name", length = 50)
    private String birdName;

    @Column(name = "bird_species", length = 50)
    private String birdSpecies;

    @Column(name = "lost_location", length = 200)
    private String lostLocation;

    @Column(name = "contact_phone", length = 20)
    private String contactPhone;

    @Column(length = 50)
    private String reward;

    @Column(name = "is_found")
    private Boolean isFound = false;

    @OneToMany(mappedBy = "post", cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderBy("sortOrder ASC")
    private List<PostImage> images = new ArrayList<>();

    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    public ForumPost() {}

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    public User getAuthor() { return author; }
    public void setAuthor(User author) { this.author = author; }
    public String getContent() { return content; }
    public void setContent(String content) { this.content = content; }
    public String getPostType() { return postType; }
    public void setPostType(String postType) { this.postType = postType; }
    public String getMediaType() { return mediaType; }
    public void setMediaType(String mediaType) { this.mediaType = mediaType; }
    public String getVideoUrl() { return videoUrl; }
    public void setVideoUrl(String videoUrl) { this.videoUrl = videoUrl; }
    public String getVideoCover() { return videoCover; }
    public void setVideoCover(String videoCover) { this.videoCover = videoCover; }
    public Integer getVideoDuration() { return videoDuration; }
    public void setVideoDuration(Integer videoDuration) { this.videoDuration = videoDuration; }
    public Integer getLikeCount() { return likeCount; }
    public void setLikeCount(Integer likeCount) { this.likeCount = likeCount; }
    public Integer getCommentCount() { return commentCount; }
    public void setCommentCount(Integer commentCount) { this.commentCount = commentCount; }
    public Integer getViewCount() { return viewCount; }
    public void setViewCount(Integer viewCount) { this.viewCount = viewCount; }
    public BigDecimal getLatitude() { return latitude; }
    public void setLatitude(BigDecimal latitude) { this.latitude = latitude; }
    public BigDecimal getLongitude() { return longitude; }
    public void setLongitude(BigDecimal longitude) { this.longitude = longitude; }
    public String getLocationName() { return locationName; }
    public void setLocationName(String locationName) { this.locationName = locationName; }
    public String getBirdName() { return birdName; }
    public void setBirdName(String birdName) { this.birdName = birdName; }
    public String getBirdSpecies() { return birdSpecies; }
    public void setBirdSpecies(String birdSpecies) { this.birdSpecies = birdSpecies; }
    public String getLostLocation() { return lostLocation; }
    public void setLostLocation(String lostLocation) { this.lostLocation = lostLocation; }
    public String getContactPhone() { return contactPhone; }
    public void setContactPhone(String contactPhone) { this.contactPhone = contactPhone; }
    public String getReward() { return reward; }
    public void setReward(String reward) { this.reward = reward; }
    public Boolean getIsFound() { return isFound; }
    public void setIsFound(Boolean isFound) { this.isFound = isFound; }
    public List<PostImage> getImages() { return images; }
    public void setImages(List<PostImage> images) { this.images = images; }
    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }
    public LocalDateTime getUpdatedAt() { return updatedAt; }
    public void setUpdatedAt(LocalDateTime updatedAt) { this.updatedAt = updatedAt; }

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
        if (likeCount == null) likeCount = 0;
        if (commentCount == null) commentCount = 0;
        if (viewCount == null) viewCount = 0;
        if (isFound == null) isFound = false;
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: backend-swift-latest/Sources/App/Models/ForumPost.swift
import Fluent
import Vapor

/// 论坛帖子模型
final class ForumPost: Model, Content, @unchecked Sendable {
    static let schema = "forum_posts"
    
    @ID(custom: "id", generatedBy: .database)
    var id: Int64?
    
    @Field(key: "author_id")
    var authorId: Int64
    
    @Field(key: "content")
    var content: String
    
    @OptionalField(key: "post_type")
    var postType: String?
    
    @OptionalField(key: "media_type")
    var mediaType: String?
    
    @OptionalField(key: "video_url")
    var videoUrl: String?
    
    @OptionalField(key: "video_cover")
    var videoCover: String?
    
    @OptionalField(key: "video_duration")
    var videoDuration: Int?
    
    @Field(key: "like_count")
    var likeCount: Int
    
    @Field(key: "comment_count")
    var commentCount: Int
    
    @Field(key: "view_count")
    var viewCount: Int
    
    @OptionalField(key: "latitude")
    var latitude: Double?
    
    @OptionalField(key: "longitude")
    var longitude: Double?
    
    @OptionalField(key: "location_name")
    var locationName: String?
    
    @OptionalField(key: "bird_id")
    var birdId: Int64?
    
    @OptionalField(key: "bird_ids")
    var birdIds: String?
    
    @OptionalField(key: "birds_info")
    var birdsInfo: String?
    
    @OptionalField(key: "bird_name")
    var birdName: String?
    
    @OptionalField(key: "bird_species")
    var birdSpecies: String?
    
    @OptionalField(key: "bird_avatar")
    var birdAvatar: String?
    
    @OptionalField(key: "lost_location")
    var lostLocation: String?
    
    @OptionalField(key: "contact_phone")
    var contactPhone: String?
    
    @OptionalField(key: "reward")
    var reward: String?
    
    @OptionalField(key: "is_found")
    var isFound: Bool?
    
    @Timestamp(key: "created_at", on: .create)
    var createdAt: Date?
    
    @Timestamp(key: "updated_at", on: .update)
    var updatedAt: Date?
    
    init() {
        self.postType = "NORMAL"
        self.mediaType = "IMAGE"
        self.likeCount = 0
        self.commentCount = 0
        self.viewCount = 0
        self.isFound = false
    }
    
    init(
        id: Int64? = nil,
        authorId: Int64,
        content: String,
        postType: String = "NORMAL",
        mediaType: String = "IMAGE",
        videoUrl: String? = nil,
        videoCover: String? = nil,
        videoDuration: Int? = nil,
        latitude: Double? = nil,
        longitude: Double? = nil,
        locationName: String? = nil,
        birdId: Int64? = nil,
        birdIds: String? = nil,
        birdsInfo: String? = nil,
        birdName: String? = nil,
        birdSpecies: String? = nil,
        birdAvatar: String? = nil,
        lostLocation: String? = nil,
        contactPhone: String? = nil,
        reward: String? = nil
    ) {
        self.id = id
        self.authorId = authorId
        self.content = content
        self.postType = postType
        self.mediaType = mediaType
        self.videoUrl = videoUrl
        self.videoCover = videoCover
        self.videoDuration = videoDuration
        self.likeCount = 0
        self.commentCount = 0
        self.viewCount = 0
        self.latitude = latitude
        self.longitude = longitude
        self.locationName = locationName
        self.birdId = birdId
        self.birdIds = birdIds
        self.birdsInfo = birdsInfo
        self.birdName = birdName
        self.birdSpecies = birdSpecies
        self.birdAvatar = birdAvatar
        self.lostLocation = lostLocation
        self.contactPhone = contactPhone
        self.reward = reward
        self.isFound = false
    }
}

// MARK: - 数据库迁移
struct CreateForumPost: AsyncMigration {
    func prepare(on database: Database) async throws {
        try await database.schema(ForumPost.schema)
            .field("id", .int64, .identifier(auto: true))
            .field("author_id", .int64, .required, .references(User.schema, "id", onDelete: .cascade))
            .field("content", .sql(raw: "TEXT"), .required)
            .field("post_type", .string, .required, .sql(.default("NORMAL")))
            .field("media_type", .string, .required, .sql(.default("IMAGE")))
            .field("video_url", .string)
            .field("video_cover", .string)
            .field("video_duration", .int)
            .field("like_count", .int, .required, .sql(.default(0)))
            .field("comment_count", .int, .required, .sql(.default(0)))
            .field("view_count", .int, .required, .sql(.default(0)))
            .field("latitude", .double)
            .field("longitude", .double)
            .field("location_name", .string)
            .field("bird_id", .int64)
            .field("bird_ids", .string)
            .field("birds_info", .string)
            .field("bird_name", .string)
            .field("bird_species", .string)
            .field("bird_avatar", .string)
            .field("lost_location", .string)
            .field("contact_phone", .string)
            .field("reward", .string)
            .field("is_found", .bool, .required, .sql(.default(false)))
            .field("created_at", .datetime)
            .field("updated_at", .datetime)
            .create()
    }
    
    func revert(on database: Database) async throws {
        try await database.schema(ForumPost.schema).delete()
    }
}
